### PR TITLE
Add parametric tests for the W3C Tracecontext RFC

### PR DESCRIPTION
**Pull request by zacharycmontoya**
## Description

Implements configuration/input header parametric tests for the following propagators:
- `none`
- `b3multi` (same as our current `b3` propagator)
- `b3 single header` (a.k.a. OTEL's "b3" format)
- `Datadog`
- `tracecontext`
- Precedence of multiple propagators like `Datadog,tracecontext`, `tracecontext,Datadog`, etc.

Adds additional traceparent/tracestate unit tests from the w3c tracecontext repo, including their Tracestate python class.

## Check list

Your PR is not ready to be reviewed? Please save it as a draft :pray:

- [ ] Follows the style guidelines of this project (See [how to easily lint the code](https://github.com/DataDog/system-tests/blob/main/docs/edit/lint.md))
- [ ] CI is passing
- [ ] There is at least one approval from code owners

Yes to all? Feel free to merge it whenever you want :heart:

### CODE REVIEW

File: parametric/spec/tracecontext.py
Comment (zacharycmontoya) [inline parametric/spec/tracecontext.py:1]: I'm not sure if this license allows us to include the source code or not. The other OpenTelemetry SDK repos do a `git clone` on the `w3c/tracecontext` repo and run these tests by implementing an HTTP server. That can be a fallback if we are unable to include the source in this repo.

File: parametric/utils/headers.py
Comment (Kyle-Verhoog) [inline parametric/utils/headers.py:4]:
```suggestion
def make_single_request_and_get_inject_headers(test_library, headers_list):
```

maybe a tad more clear?

**Review by Kyle-Verhoog — COMMENTED** [review]
Didn't review every single test case, but the high-level setup and organization looks good to me! Great stuff @zacharycmontoya 👏

File: parametric/spec/tracecontext.py
Comment (zacharycmontoya) [reply parametric/spec/tracecontext.py]: This is resolved now after receiving guidance from colleagues

File: parametric/test_headers_tracecontext.py
Comment (mabdinur) [inline parametric/test_headers_tracecontext.py:829]: According to the w3c spec tabs are invalid value: https://www.w3.org/TR/trace-context/#value. I think we need use another character in this test

Comment (zacharycmontoya) [reply parametric/test_headers_tracecontext.py:829]: This test case came from the w3c/tracecontext so it should be valid 😆

And looking at [this grammar](https://www.w3.org/TR/trace-context/#list), it does appear to be valid because:
1. OWS (optional whitespace) is allowed between the previous comma and the next `key=value`
2. OWS (optional whitespace) is allowed after a `key=value` and before the next comma

Comment (zacharycmontoya) [reply parametric/test_headers_tracecontext.py:829]: Also [in this section](https://www.w3.org/TR/trace-context/#tracestate-header-field-values) it asserts the following:

> Spaces and horizontal tabs surrounding list-members are ignored.

I think what would be invalid was if there was a \t in the middle of a value, like `foo=123\t456`

**Comment by mabdinur** [comment]
Failing w3c python tests should be fixed by the following PR: https://github.com/DataDog/dd-trace-py/pull/4791

File: parametric/utils/test_agent.py
Comment (zacharycmontoya) [reply parametric/utils/test_agent.py]: Thanks, done in fc47b9ff8ddcd4e1fb072f5ca73a294e261594a0

File: parametric/utils/headers.py
Comment (zacharycmontoya) [reply parametric/utils/headers.py]: Good suggestion, implemented in d45c213f8ba18734c1e28faa30ff220956726e7e

File: parametric/test_headers_tracestate_dd.py
Comment (ZStriker19) [inline parametric/test_headers_tracestate_dd.py:289]: Wouldn't this be the correct encoding? Changing `=` to `~` rather than `_` ?

File: parametric/test_headers_tracecontext.py
Comment (zacharycmontoya) [reply parametric/test_headers_tracecontext.py:817]: Resolving, but let me know if you have further questions

File: parametric/test_headers_tracestate_dd.py
Comment (zacharycmontoya) [reply parametric/test_headers_tracestate_dd.py:289]: Thanks for this fix, this makes sense to standardize string injection/extraction

File: parametric/test_headers_precedence.py
Comment (mabdinur) [inline parametric/test_headers_precedence.py:34]: Python should now enable w3c by default. Is this test still failing?

File: parametric/test_headers_tracecontext.py
Comment (mabdinur) [inline parametric/test_headers_tracecontext.py:779]: @ZStriker19 is this resolved? If not, we should create a jira card for this

File: parametric/test_headers_tracestate_dd.py
Comment (mabdinur) [inline parametric/test_headers_tracestate_dd.py:326]: I believe the issue with python was the `=` character was not encoded correctly (parsing header2 raised a syntax error).  I believe this bug was fixed. 

Is this test passing now?

File: parametric/test_headers_b3.py
Comment (mabdinur) [inline parametric/test_headers_b3.py:59]: @ZStriker19 Is this still an issue

File: parametric/spec/tracecontext.py
Comment (mabdinur) [inline parametric/spec/tracecontext.py:123]: nit: I don't think we need to define/use `to_string`.   We can just overload the built in `__str__` method.

**Review by mabdinur — COMMENTED** [review]
left some questions to the team. Overal looks good to me

File: parametric/test_headers_b3.py
Comment (zacharycmontoya) [reply parametric/test_headers_b3.py:59]: I added this skip very recently when because I noticed this test failure, so I think it is still an issue that will need to be addressed

File: parametric/test_headers_precedence.py
Comment (zacharycmontoya) [reply parametric/test_headers_precedence.py]: I just re-enabled this test case and it's passing! Resolving

File: parametric/test_headers_tracestate_dd.py
Comment (zacharycmontoya) [reply parametric/test_headers_tracestate_dd.py]: I just re-enabled the test case and it's passing now 👍🏼 Resolving

File: parametric/spec/tracecontext.py
Comment (zacharycmontoya) [reply parametric/spec/tracecontext.py:123]: `to_string` is a helper function that `__str__` and other methods call into. I'm okay either way, but it seems fine to me as-is

**Review by Qard — APPROVED** [review]

**Comment by dianashevchenko** [comment]
Have a small nit for `test_headers_tracestate_dd_propagate_origin` headers_3, but I'll leave it at your discretion

```
 # 3) x-datadog-origin has invalid characters. Since tilde must be unescaped during extraction,
    # all invalid characters including '~', must be replaced with '_',
    # and after that '=' must be replaced with `~`
    # Result: Origin set to header value, where invalid characters replaced by '_'
```

In such case, `synthetics~,=web` will become `synthetics__~web` and **not** `synthetics~_~web`

File: parametric/test_headers_b3.py
Comment (ZStriker19) [reply parametric/test_headers_b3.py:59]: I'll make the fix.

File: parametric/test_span_sampling.py
Comment (Kyle-Verhoog) [inline parametric/test_span_sampling.py:351]: Unrelated? I don't see it failing on `main`. Or was this flaky?

File: parametric/utils/test_agent.py
Comment (Kyle-Verhoog) [inline parametric/utils/test_agent.py:1]: @zacharycmontoya can we add this to the test_agent fixture (`_TestAgentAPI`) for convenience?

I'm ok holding off for now and refactoring it afterwards since this is just a nice-to-have.

File: parametric/README.md
Comment (Kyle-Verhoog) [inline parametric/README.md:39]: 👍 thanks for adding this

**Review by Kyle-Verhoog — APPROVED** [review]
A few comments, overall this is great! I spot checked a few tests and they look good.

The suggestion we can probably refactor after this is merged.

Great stuff everyone 👏 👏 👏

File: parametric/test_headers_b3.py
Comment (ZStriker19) [reply parametric/test_headers_b3.py:59]: https://github.com/DataDog/dd-trace-py/pull/4976

File: parametric/test_span_sampling.py
Comment (ZStriker19) [inline parametric/test_span_sampling.py:351]: nit: should this be a separate PR?

File: parametric/test_tracer.py
Comment (ZStriker19) [inline parametric/test_tracer.py:64]: nit: should this also be a separate PR?

File: parametric/test_span_sampling.py
Comment (zacharycmontoya) [reply parametric/test_span_sampling.py:351]: Yes it was flaky so I figured it was best to skip this for now to make sure I could get a clean test `parametric (nodejs)` test run

Comment (zacharycmontoya) [reply parametric/test_span_sampling.py:351]: This was flaky, so I skipped

File: parametric/test_tracer.py
Comment (zacharycmontoya) [reply parametric/test_tracer.py:64]: This was flaky, so I skipped

File: parametric/test_headers_b3.py
Comment (zacharycmontoya) [reply parametric/test_headers_b3.py:59]: We can unskip the test in a separate PR later so that we can merge this large set of changes right away

File: parametric/test_span_sampling.py
Comment (Kyle-Verhoog) [reply parametric/test_span_sampling.py:351]: Can we update the `reason` to be that the test is flaky rather than not implemented?

Comment (Qard) [reply parametric/test_span_sampling.py:351]: Possibly a flake or local misconfiguration? I don't recall running into any failures here in my local testing. 🤷🏻

**Review by cbeauchesne — APPROVED** [review]

File: parametric/test_span_sampling.py
Comment (zacharycmontoya) [reply parametric/test_span_sampling.py]: Yep good call. Done in aebb48100fade516821a194bf566dc835dfecc54

**Review by ZStriker19 — APPROVED** [review]

File: parametric/test_headers_precedence.py
Comment (bantonsson) [inline parametric/test_headers_precedence.py:68]: @zacharycmontoya As far as I know, the only [valid values](https://github.com/Datadog/architecture/blob/master/rfcs/apm/integrations/priority-sampling/rfc.md#sampling-values-and-propagation) for sampling priority are `-1`, `0`, `1` and `2`. Java enforces this and all these tests fail because of that. Would it be ok if I opened a PR with these set to `-1` instead?

Comment (zacharycmontoya) [reply parametric/test_headers_precedence.py:68]: Yes that sounds good 👍🏼

Comment (bantonsson) [reply parametric/test_headers_precedence.py:68]: I think we should leave it as is. It seems that the tracers should indeed accept other incoming sampling priorities, and just pass them through. I'll make Java behave the same way.

**Pull request merged**
